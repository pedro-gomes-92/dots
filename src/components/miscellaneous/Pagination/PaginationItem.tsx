import React, { FunctionComponent } from 'react';
import classnames from 'classnames';
import { TextButton, TextButtonProps } from '../../actions';

export interface PaginationItemProps extends TextButtonProps {
  text: string;
  isCurrent?: boolean;
}

export const PaginationItem: FunctionComponent<PaginationItemProps> = ({
  className,
  isCurrent,
  ...rest
}: PaginationItemProps): JSX.Element => (
  <TextButton className={classnames('pagination-link', { 'is-current': isCurrent }, className)} {...rest} />
);

PaginationItem.defaultProps = {
  isCurrent: false,
};
