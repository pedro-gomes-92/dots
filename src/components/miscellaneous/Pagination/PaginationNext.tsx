import React, { FunctionComponent } from 'react';
import classnames from 'classnames';
import { ButtonProps, Button } from '../../actions';

export interface PaginationNextProps extends ButtonProps {}

export const PaginationNext: FunctionComponent<PaginationNextProps> = ({
  className,
  ...rest
}: PaginationNextProps): JSX.Element => <Button className={classnames('pagination-next', className)} {...rest} />;

PaginationNext.defaultProps = {
  icon: 'chevron-right-outline',
};
