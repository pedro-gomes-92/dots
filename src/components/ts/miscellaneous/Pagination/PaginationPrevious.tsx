import React, { FunctionComponent } from 'react';
import classnames from 'classnames';
import { ButtonProps, Button } from '../../actions';

export interface PaginationPreviousProps extends ButtonProps {}

export const PaginationPrevious: FunctionComponent<PaginationPreviousProps> = ({
  className,
  ...rest
}: PaginationPreviousProps): JSX.Element => (
  <Button className={classnames('pagination-previous', className)} {...rest} />
);
PaginationPrevious.defaultProps = {
  icon: 'chevron-left-outline',
};
