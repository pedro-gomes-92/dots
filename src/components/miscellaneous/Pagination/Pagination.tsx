import React, { FunctionComponent, useState } from 'react';
import classnames from 'classnames';
import { BaseProps, Base } from '../../Base';
import { PaginationItem } from './PaginationItem';
import { PaginationPrevious } from './PaginationPrevious';
import { PaginationNext } from './PaginationNext';
import { List, ListItem } from '../../data';

export interface PaginationProps extends BaseProps {
  total: number;
  range?: number;
  onPageChange?: (newPage: number) => void;
}

export const Pagination: FunctionComponent<PaginationProps> = ({
  className,
  total,
  range,
  onPageChange,
  ...rest
}: PaginationProps): JSX.Element => {
  const [currentPage, setCurrentPage] = useState(1);

  const handlePageChange = (newPage: number): void => {
    setCurrentPage(newPage);
    onPageChange(newPage);
  };

  const renderItem = (pageNumber: number, text: string): JSX.Element => {
    return (
      <ListItem key={`list-item-${pageNumber - 1}`}>
        <PaginationItem
          text={text}
          isCurrent={currentPage === pageNumber}
          onClick={(): void => {
            handlePageChange(pageNumber);
          }}
        />
      </ListItem>
    );
  };

  const items = [renderItem(1, '1')];
  const firstBoundaryStart = 2;
  const lastBoundaryEnd = total - 1;
  const lastBoundaryStart = Math.max(lastBoundaryEnd - 2 * range - 1, firstBoundaryStart);
  const firstBoundaryEnd = Math.min(lastBoundaryEnd, firstBoundaryStart + 2 * range + 1);

  let boundaryStart = Math.min(Math.max(firstBoundaryStart, currentPage - range), lastBoundaryStart);
  let boundaryEnd = Math.max(Math.min(currentPage + range, lastBoundaryEnd), firstBoundaryEnd);

  if (boundaryStart > firstBoundaryStart) {
    // Render left ellipsis
    items.push(renderItem(boundaryStart - 1, '...'));
  }

  for (let pageNumber = boundaryStart; pageNumber <= boundaryEnd; pageNumber++) {
    // Render boundary pages
    items.push(renderItem(pageNumber, `${pageNumber}`));
  }

  if (boundaryEnd < lastBoundaryEnd) {
    // Render right ellipsis
    items.push(renderItem(boundaryEnd + 1, '...'));
  }

  if (total > 1) {
    // Render last page
    items.push(renderItem(total, `${total}`));
  }

  return (
    <Base
      className={classnames('pagination', 'is-centered', className)}
      attribute={{ role: 'navigation', 'aria-label': 'pagination' }}
      {...rest}
    >
      <PaginationPrevious
        onClick={(): void => {
          handlePageChange(currentPage - 1);
        }}
        isEnabled={currentPage > 1}
      />

      <PaginationNext
        onClick={(): void => {
          handlePageChange(currentPage + 1);
        }}
        isEnabled={currentPage < total}
      />

      <List className="pagination-list">{items}</List>
    </Base>
  );
};

Pagination.defaultProps = {
  tag: 'nav',
  range: 2,
  onPageChange: (): void => {},
};
