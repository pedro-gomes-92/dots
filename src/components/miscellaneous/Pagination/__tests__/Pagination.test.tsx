import React from 'react';
import { render, fireEvent } from '@testing-library/react';

import { Pagination } from '../Pagination';

describe('Given <Pagination />', (): void => {
  let container: HTMLElement;
  const getList = (html: HTMLElement): HTMLCollection => html.querySelector('ul.list').children;
  const getPage = (html: HTMLElement, index: number): HTMLElement =>
    getList(html)
      .item(index)
      .querySelector('.pagination-link');

  describe('When initializes', (): void => {
    it('Then should match the snapshot', (): void => {
      container = render(<Pagination total={20} />).container;
      expect(container).toMatchSnapshot();
    });

    describe('And uses total', (): void => {
      it('Then should render total pages', (): void => {
        container = render(<Pagination total={4} />).container;
        expect(getList(container).length).toEqual(4);
      });
    });

    /*
    describe('And uses invalid total', (): void => {
      it('Then should throw an error', (): void => {
        const invalid = (): void => {
          render(<Pagination total={0} />);
        };

        expect(invalid).toThrowError('Invalid values on properties');
      });
    });
    */

    describe('And uses range', (): void => {
      beforeEach((): void => {
        container = render(<Pagination total={20} range={2} />).container;
      });

      it('Then should match the snapshot', (): void => {
        expect(container).toMatchSnapshot();
      });

      it('And should render no more than ranged pages', (): void => {
        const list = getList(container);
        expect(list.length).toEqual(9);

        fireEvent.click(getPage(container, 7));

        expect(list.item(1).querySelector('.text').textContent).toEqual('...');
        expect(list.item(7).querySelector('.text').textContent).toEqual('...');
      });
    });

    /* 
    describe('And uses invalid range', (): void => {
      it('Then should throw an error', (): void => {
        const invalid = (): void => {
          render(<Pagination total={20} range={-1} />);
        };

        expect(invalid).toThrowError('Invalid values on properties');
      });
    });
    */
  });

  describe('When changes page', (): void => {
    it('Then should trigger onPageChange', (): void => {
      const pageChange = jest.fn();
      container = render(<Pagination total={20} onPageChange={pageChange} />).container;
      fireEvent.click(getPage(container, 7));

      expect(pageChange).toBeCalledTimes(1);
    });
  });
});
