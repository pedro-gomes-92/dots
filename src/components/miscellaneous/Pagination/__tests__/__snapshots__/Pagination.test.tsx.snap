// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Given <Pagination /> When initializes And uses range Then should match the snapshot 1`] = `
<div>
  <nav
    aria-label="pagination"
    class="pagination is-centered"
    role="navigation"
  >
    <button
      class="is-disabled button pagination-previous"
      disabled=""
    >
      <span
        class="icon"
      >
        <i
          class="icon-font icon-font-chevron-left-outline"
        />
      </span>
    </button>
    <button
      class="button pagination-next"
    >
      <span
        class="icon"
      >
        <i
          class="icon-font icon-font-chevron-right-outline"
        />
      </span>
    </button>
    <ul
      class="list pagination-list"
    >
      <li
        class="list-item"
      >
        <button
          class="button is-text pagination-link is-current"
        >
          <span
            class="text"
          >
            1
          </span>
        </button>
      </li>
      <li
        class="list-item"
      >
        <button
          class="button is-text pagination-link"
        >
          <span
            class="text"
          >
            2
          </span>
        </button>
      </li>
      <li
        class="list-item"
      >
        <button
          class="button is-text pagination-link"
        >
          <span
            class="text"
          >
            3
          </span>
        </button>
      </li>
      <li
        class="list-item"
      >
        <button
          class="button is-text pagination-link"
        >
          <span
            class="text"
          >
            4
          </span>
        </button>
      </li>
      <li
        class="list-item"
      >
        <button
          class="button is-text pagination-link"
        >
          <span
            class="text"
          >
            5
          </span>
        </button>
      </li>
      <li
        class="list-item"
      >
        <button
          class="button is-text pagination-link"
        >
          <span
            class="text"
          >
            6
          </span>
        </button>
      </li>
      <li
        class="list-item"
      >
        <button
          class="button is-text pagination-link"
        >
          <span
            class="text"
          >
            7
          </span>
        </button>
      </li>
      <li
        class="list-item"
      >
        <button
          class="button is-text pagination-link"
        >
          <span
            class="text"
          >
            ...
          </span>
        </button>
      </li>
      <li
        class="list-item"
      >
        <button
          class="button is-text pagination-link"
        >
          <span
            class="text"
          >
            20
          </span>
        </button>
      </li>
    </ul>
  </nav>
</div>
`;

exports[`Given <Pagination /> When initializes Then should match the snapshot 1`] = `
<div>
  <nav
    aria-label="pagination"
    class="pagination is-centered"
    role="navigation"
  >
    <button
      class="is-disabled button pagination-previous"
      disabled=""
    >
      <span
        class="icon"
      >
        <i
          class="icon-font icon-font-chevron-left-outline"
        />
      </span>
    </button>
    <button
      class="button pagination-next"
    >
      <span
        class="icon"
      >
        <i
          class="icon-font icon-font-chevron-right-outline"
        />
      </span>
    </button>
    <ul
      class="list pagination-list"
    >
      <li
        class="list-item"
      >
        <button
          class="button is-text pagination-link is-current"
        >
          <span
            class="text"
          >
            1
          </span>
        </button>
      </li>
      <li
        class="list-item"
      >
        <button
          class="button is-text pagination-link"
        >
          <span
            class="text"
          >
            2
          </span>
        </button>
      </li>
      <li
        class="list-item"
      >
        <button
          class="button is-text pagination-link"
        >
          <span
            class="text"
          >
            3
          </span>
        </button>
      </li>
      <li
        class="list-item"
      >
        <button
          class="button is-text pagination-link"
        >
          <span
            class="text"
          >
            4
          </span>
        </button>
      </li>
      <li
        class="list-item"
      >
        <button
          class="button is-text pagination-link"
        >
          <span
            class="text"
          >
            5
          </span>
        </button>
      </li>
      <li
        class="list-item"
      >
        <button
          class="button is-text pagination-link"
        >
          <span
            class="text"
          >
            6
          </span>
        </button>
      </li>
      <li
        class="list-item"
      >
        <button
          class="button is-text pagination-link"
        >
          <span
            class="text"
          >
            7
          </span>
        </button>
      </li>
      <li
        class="list-item"
      >
        <button
          class="button is-text pagination-link"
        >
          <span
            class="text"
          >
            ...
          </span>
        </button>
      </li>
      <li
        class="list-item"
      >
        <button
          class="button is-text pagination-link"
        >
          <span
            class="text"
          >
            20
          </span>
        </button>
      </li>
    </ul>
  </nav>
</div>
`;
